aten/src/ATen/TensorImpl.cpp

```diff
@@ -55,13 +56,29 @@ void Tensor::backward(
   pImpl->backward(std::move(gradient), keep_graph, create_graph);
 }
 
-TensorImpl::TensorImpl(Backend backend, ScalarType scalar_type) {
-  backend_ = backend;
-  scalar_type_ = scalar_type;
+TensorImpl::TensorImpl(Backend backend, ScalarType scalar_type, bool is_variable) {
   auto type = &globalContext().getType(backend, scalar_type);
   Storage* storage = type->storage(true).release();
   StorageImpl* storage_impl = storage->pImpl();
   tensor = new THTensor(storage_impl);
+  tensor->backend_ = backend;
+  tensor->scalar_type_ = scalar_type;
+  tensor->is_variable_ = is_variable;
+}
+
+TensorImpl::TensorImpl(
+    Backend backend,
+    ScalarType scalar_type,
+    THTensor* tensor_,
+    bool is_variable) {
+  if (tensor_) {
+    tensor = tensor_;
+  } else {
+    tensor = new THTensor(nullptr);
```

Comment:
when does this case come up?  For sparse tensors?  Why did you choose this over having a null tensor (which is what we used to have, right?).  Not having a storage doesn't look like a good tradeoff.

Reply:
Basically, the surface area of TensorImpl is a lot smaller than the surface area for THTensor; it's a lot easier to handle null tensor_ than to handle null storage.

Reply:
You need to store scalar_type_ and backend_ somewhere. Since it's a field it'll be part of THTensor. Even if there's not storage behind THTensor, the object still needs to exist to support functions like 'type()'.

Reply:
Is it correct that this only happens with sparse tensors?

Reply:
``` 
aten/src/ATen/templates/TypeDerived.cpp:99:
Tensor ${Type}::unsafeTensorFromTH could technically use the constructor with a nullptr argument

aten/src/ATen/SparseTensorImpl.cpp:22 explicitly calls the constructor with a nullptr

torch/csrc/jit/interpreter.cpp:340 calls it with TensorImpl(at::Backend::Undefined,at::ScalarType::Undefined, nullptr, /* is_variable */ false)
to construct ContainerTensor

torch/csrc/autograd/variable.cpp:26
Variable::Impl::Impl calls into it with
TensorImpl(data.type().backend(), data.type().scalarType(), nullptr, /* is variable */ true)
for construction and also has a separate data_ field for a Tensor
```

Reply:
So it's actually used a lot